# Final Project - Beat the Bookies

## Day One - Initial Data Exploration

The goal of today is:

- Create repositories
- Add collaborators
- Add notebooks to the project
- Confirm primary data source
- Build a dataframe that includes all the relevant and desired data from our model (check google sheets for the columns). 


Using the dataset downloaded from kaggle (https://www.kaggle.com/hugomathien/soccer?) - exploring the data available for one season (2015/16) to combine the data into our own dataframe that includes the relevant information that we need for our models.

Colummns for the dataset:

- Home Team Rank
- Home Team Total Wins
- Home Team win
- Draw
- Away Team Win
- Away Team Rank

To add columns including possesion, reds, yellows, corners, shots at goal, shots on target, goal percentage, throw ins (?), fouls, free kicks. Later on when we know that we can combine the core of the table easily for all seasons.

Target:
- Home Team win y/n

### imports 

In [1]:
import pandas as pd
import numpy as np 
import sqlite3

### importing the data from sqlite

In [2]:
# create the file path (include two beatthebookies)
path = '/Users/georgebrockman/code/georgebrockman/beatthebookies/beatthebookies/data/database.sqlite'
# create the connection
conn = sqlite3.connect(path)
# create the cursor
cursor = conn.cursor()

In [3]:
# from the match table we want to select all matches played in the English Premier league for the 2015/16 season.
# Premier League - league_id = 1729
# order by stages (Game Weeks)
# need quotes around '2015/2016'
result = cursor.execute(""" SELECT * FROM "Match" m 
                        WHERE league_id = 1729 
                        AND season = '2015/2016' 
                        ORDER BY stage ASC""")

In [4]:
row = result.fetchall()
row[0] # seems to be a lot of null values for shots, s

(4389,
 1729,
 1729,
 '2015/2016',
 1,
 '2015-08-09 00:00:00',
 1987032,
 9825,
 8654,
 0,
 2,
 1,
 2,
 4,
 6,
 8,
 4,
 6,
 3,
 5,
 7,
 5,
 1,
 2,
 4,
 6,
 8,
 3,
 5,
 7,
 5,
 4,
 6,
 1,
 3,
 3,
 3,
 3,
 6,
 6,
 8,
 8,
 8,
 11,
 1,
 3,
 3,
 3,
 3,
 6,
 6,
 6,
 8,
 10,
 10,
 30859,
 26154,
 35606,
 46539,
 38521,
 159594,
 37436,
 75489,
 36378,
 196386,
 46469,
 169756,
 109897,
 35110,
 49543,
 155782,
 37169,
 575789,
 148302,
 25496,
 18506,
 192899,
 '<goal><value><comment>n</comment><stats><goals>1</goals><shoton>1</shoton></stats><event_incident_typefk>406</event_incident_typefk><coordinates><value>21</value><value>62</value></coordinates><elapsed>43</elapsed><player2>25496</player2><subtype>header</subtype><player1>148302</player1><sortorder>2</sortorder><team>8654</team><id>4659387</id><n>251</n><type>goal</type><goal_type>n</goal_type></value><value><comment>n</comment><stats><goals>1</goals><shoton>1</shoton></stats><event_incident_typefk>39</event_incident_typefk><coordinate

In [5]:
# create a DataFrame from the SQL query
matches_df = pd.read_sql(""" SELECT * FROM "Match" m 
                        WHERE league_id = 1729 
                        AND season = '2015/2016' 
                        ORDER BY stage ASC""", conn)

In [6]:
# inspect the DataFrame
matches_df.columns.values

array(['id', 'country_id', 'league_id', 'season', 'stage', 'date',
       'match_api_id', 'home_team_api_id', 'away_team_api_id',
       'home_team_goal', 'away_team_goal', 'home_player_X1',
       'home_player_X2', 'home_player_X3', 'home_player_X4',
       'home_player_X5', 'home_player_X6', 'home_player_X7',
       'home_player_X8', 'home_player_X9', 'home_player_X10',
       'home_player_X11', 'away_player_X1', 'away_player_X2',
       'away_player_X3', 'away_player_X4', 'away_player_X5',
       'away_player_X6', 'away_player_X7', 'away_player_X8',
       'away_player_X9', 'away_player_X10', 'away_player_X11',
       'home_player_Y1', 'home_player_Y2', 'home_player_Y3',
       'home_player_Y4', 'home_player_Y5', 'home_player_Y6',
       'home_player_Y7', 'home_player_Y8', 'home_player_Y9',
       'home_player_Y10', 'home_player_Y11', 'away_player_Y1',
       'away_player_Y2', 'away_player_Y3', 'away_player_Y4',
       'away_player_Y5', 'away_player_Y6', 'away_player_Y7',
       'aw

In [7]:
# we can remove all the ones not needed for home-away goal data
simple_matches = matches_df.drop(['home_player_X1',
       'home_player_X2', 'home_player_X3', 'home_player_X4',
       'home_player_X5', 'home_player_X6', 'home_player_X7',
       'home_player_X8', 'home_player_X9', 'home_player_X10',
       'home_player_X11', 'away_player_X1', 'away_player_X2',
       'away_player_X3', 'away_player_X4', 'away_player_X5',
       'away_player_X6', 'away_player_X7', 'away_player_X8',
       'away_player_X9', 'away_player_X10', 'away_player_X11',
       'home_player_Y1', 'home_player_Y2', 'home_player_Y3',
       'home_player_Y4', 'home_player_Y5', 'home_player_Y6',
       'home_player_Y7', 'home_player_Y8', 'home_player_Y9',
       'home_player_Y10', 'home_player_Y11', 'away_player_Y1',
       'away_player_Y2', 'away_player_Y3', 'away_player_Y4',
       'away_player_Y5', 'away_player_Y6', 'away_player_Y7',
       'away_player_Y8', 'away_player_Y9', 'away_player_Y10',
       'away_player_Y11', 'home_player_1', 'home_player_2',
       'home_player_3', 'home_player_4', 'home_player_5', 'home_player_6',
       'home_player_7', 'home_player_8', 'home_player_9',
       'home_player_10', 'home_player_11', 'away_player_1',
       'away_player_2', 'away_player_3', 'away_player_4', 'away_player_5',
       'away_player_6', 'away_player_7', 'away_player_8', 'away_player_9',
       'away_player_10', 'away_player_11', 'goal', 'shoton', 'shotoff',
       'foulcommit', 'card', 'cross', 'corner', 'possession', 'B365H',
       'B365D', 'B365A', 'BWH', 'BWD', 'BWA', 'IWH', 'IWD', 'IWA', 'LBH',
       'LBD', 'LBA', 'PSH', 'PSD', 'PSA', 'WHH', 'WHD', 'WHA', 'SJH',
       'SJD', 'SJA', 'VCH', 'VCD', 'VCA', 'GBH', 'GBD', 'GBA', 'BSH',
       'BSD', 'BSA'], axis=1)

In [8]:
simple_matches.head(5)

id  country_id  league_id     season  stage                 date  \
0  4389        1729       1729  2015/2016      1  2015-08-09 00:00:00   
1  4390        1729       1729  2015/2016      1  2015-08-08 00:00:00   
2  4391        1729       1729  2015/2016      1  2015-08-08 00:00:00   
3  4392        1729       1729  2015/2016      1  2015-08-08 00:00:00   
4  4393        1729       1729  2015/2016      1  2015-08-08 00:00:00   

   match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
0       1987032              9825              8654               0   
1       1987033              8678             10252               0   
2       1987034              8455             10003               2   
3       1987035              8668              9817               2   
4       1987036              8197              8472               4   

   away_team_goal  
0               2  
1               1  
2               2  
3               2  
4               2

In [9]:
simple_matches['home_goals']= simple_matches.groupby('home_team_api_id')['home_team_goal'].cumsum().fillna(0)

In [10]:
simple_matches['away_goals']= simple_matches.groupby('away_team_api_id')['away_team_goal'].cumsum().fillna(0)

In [11]:
simple_matches

id  country_id  league_id     season  stage                 date  \
0    4389        1729       1729  2015/2016      1  2015-08-09 00:00:00   
1    4390        1729       1729  2015/2016      1  2015-08-08 00:00:00   
2    4391        1729       1729  2015/2016      1  2015-08-08 00:00:00   
3    4392        1729       1729  2015/2016      1  2015-08-08 00:00:00   
4    4393        1729       1729  2015/2016      1  2015-08-08 00:00:00   
5    4394        1729       1729  2015/2016      1  2015-08-08 00:00:00   
6    4395        1729       1729  2015/2016      1  2015-08-09 00:00:00   
7    4396        1729       1729  2015/2016      1  2015-08-08 00:00:00   
8    4397        1729       1729  2015/2016      1  2015-08-09 00:00:00   
9    4398        1729       1729  2015/2016      1  2015-08-10 00:00:00   
10   4499        1729       1729  2015/2016      2  2015-08-14 00:00:00   
11   4500        1729       1729  2015/2016      2  2015-08-16 00:00:00   
12   4501        1729       1729  2015/2016      2  2015-08-17 00:00:00   
13   4502        1729       1729  2015/2016      2  2015-08-16 00:00:00   
14   4503        1729       1729  2015/2016      2  2015-08-15 00:00:00   
15   4504        1729       1729  2015/2016      2  2015-08-15 00:00:00   
16   4505        1729       1729  2015/2016      2  2015-08-15 00:00:00   
17   4506        1729       1729  2015/2016      2  2015-08-15 00:00:00   
18   4507        1729       1729  2015/2016      2  2015-08-15 00:00:00   
19   4508        1729       1729  2015/2016      2  2015-08-15 00:00:00   
20   4609        1729       1729  2015/2016      3  2015-08-24 00:00:00   
21   4610        1729       1729  2015/2016      3  2015-08-22 00:00:00   
22   4611        1729       1729  2015/2016      3  2015-08-23 00:00:00   
23   4612        1729       1729  2015/2016      3  2015-08-22 00:00:00   
24   4613        1729       1729  2015/2016      3  2015-08-22 00:00:00   
25   4614        1729       1729  2015/2016      3  2015-08-22 00:00:00   
26   4615        1729       1729  2015/2016      3  2015-08-22 00:00:00   
27   4616        1729       1729  2015/2016      3  2015-08-23 00:00:00   
28   4617        1729       1729  2015/2016      3  2015-08-23 00:00:00   
29   4618        1729       1729  2015/2016      3  2015-08-22 00:00:00   
..    ...         ...        ...        ...    ...                  ...   
350  4679        1729       1729  2015/2016     36  2016-04-30 00:00:00   
351  4680        1729       1729  2015/2016     36  2016-05-02 00:00:00   
352  4681        1729       1729  2015/2016     36  2016-04-30 00:00:00   
353  4682        1729       1729  2015/2016     36  2016-05-01 00:00:00   
354  4683        1729       1729  2015/2016     36  2016-04-30 00:00:00   
355  4684        1729       1729  2015/2016     36  2016-05-01 00:00:00   
356  4685        1729       1729  2015/2016     36  2016-04-30 00:00:00   
357  4686        1729       1729  2015/2016     36  2016-05-01 00:00:00   
358  4687        1729       1729  2015/2016     36  2016-04-30 00:00:00   
359  4688        1729       1729  2015/2016     36  2016-04-30 00:00:00   
360  4689        1729       1729  2015/2016     37  2016-05-07 00:00:00   
361  4690        1729       1729  2015/2016     37  2016-05-07 00:00:00   
362  4691        1729       1729  2015/2016     37  2016-05-07 00:00:00   
363  4692        1729       1729  2015/2016     37  2016-05-07 00:00:00   
364  4693        1729       1729  2015/2016     37  2016-05-08 00:00:00   
365  4694        1729       1729  2015/2016     37  2016-05-08 00:00:00   
366  4695        1729       1729  2015/2016     37  2016-05-07 00:00:00   
367  4696        1729       1729  2015/2016     37  2016-05-07 00:00:00   
368  4697        1729       1729  2015/2016     37  2016-05-08 00:00:00   
369  4698        1729       1729  2015/2016     37  2016-05-07 00:00:00   
370  4699        1729       1729  2015/2016     38  2016-05-15 00:00:00   
371  4700        1729       1729

In [12]:
simple_matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 380 entries, 0 to 379
Data columns (total 13 columns):
id                  380 non-null int64
country_id          380 non-null int64
league_id           380 non-null int64
season              380 non-null object
stage               380 non-null int64
date                380 non-null object
match_api_id        380 non-null int64
home_team_api_id    380 non-null int64
away_team_api_id    380 non-null int64
home_team_goal      380 non-null int64
away_team_goal      380 non-null int64
home_goals          380 non-null int64
away_goals          380 non-null int64
dtypes: int64(11), object(2)
memory usage: 38.7+ KB


In [13]:
simple_matches['home_wins'] = simple_matches.home_team_goal > simple_matches.away_team_goal
simple_matches['draws'] = simple_matches.home_team_goal == simple_matches.away_team_goal
simple_matches['away_wins'] = simple_matches.home_team_goal < simple_matches.away_team_goal
simple_matches

id  country_id  league_id     season  stage                 date  \
0    4389        1729       1729  2015/2016      1  2015-08-09 00:00:00   
1    4390        1729       1729  2015/2016      1  2015-08-08 00:00:00   
2    4391        1729       1729  2015/2016      1  2015-08-08 00:00:00   
3    4392        1729       1729  2015/2016      1  2015-08-08 00:00:00   
4    4393        1729       1729  2015/2016      1  2015-08-08 00:00:00   
5    4394        1729       1729  2015/2016      1  2015-08-08 00:00:00   
6    4395        1729       1729  2015/2016      1  2015-08-09 00:00:00   
7    4396        1729       1729  2015/2016      1  2015-08-08 00:00:00   
8    4397        1729       1729  2015/2016      1  2015-08-09 00:00:00   
9    4398        1729       1729  2015/2016      1  2015-08-10 00:00:00   
10   4499        1729       1729  2015/2016      2  2015-08-14 00:00:00   
11   4500        1729       1729  2015/2016      2  2015-08-16 00:00:00   
12   4501        1729       1729  2015/2016      2  2015-08-17 00:00:00   
13   4502        1729       1729  2015/2016      2  2015-08-16 00:00:00   
14   4503        1729       1729  2015/2016      2  2015-08-15 00:00:00   
15   4504        1729       1729  2015/2016      2  2015-08-15 00:00:00   
16   4505        1729       1729  2015/2016      2  2015-08-15 00:00:00   
17   4506        1729       1729  2015/2016      2  2015-08-15 00:00:00   
18   4507        1729       1729  2015/2016      2  2015-08-15 00:00:00   
19   4508        1729       1729  2015/2016      2  2015-08-15 00:00:00   
20   4609        1729       1729  2015/2016      3  2015-08-24 00:00:00   
21   4610        1729       1729  2015/2016      3  2015-08-22 00:00:00   
22   4611        1729       1729  2015/2016      3  2015-08-23 00:00:00   
23   4612        1729       1729  2015/2016      3  2015-08-22 00:00:00   
24   4613        1729       1729  2015/2016      3  2015-08-22 00:00:00   
25   4614        1729       1729  2015/2016      3  2015-08-22 00:00:00   
26   4615        1729       1729  2015/2016      3  2015-08-22 00:00:00   
27   4616        1729       1729  2015/2016      3  2015-08-23 00:00:00   
28   4617        1729       1729  2015/2016      3  2015-08-23 00:00:00   
29   4618        1729       1729  2015/2016      3  2015-08-22 00:00:00   
..    ...         ...        ...        ...    ...                  ...   
350  4679        1729       1729  2015/2016     36  2016-04-30 00:00:00   
351  4680        1729       1729  2015/2016     36  2016-05-02 00:00:00   
352  4681        1729       1729  2015/2016     36  2016-04-30 00:00:00   
353  4682        1729       1729  2015/2016     36  2016-05-01 00:00:00   
354  4683        1729       1729  2015/2016     36  2016-04-30 00:00:00   
355  4684        1729       1729  2015/2016     36  2016-05-01 00:00:00   
356  4685        1729       1729  2015/2016     36  2016-04-30 00:00:00   
357  4686        1729       1729  2015/2016     36  2016-05-01 00:00:00   
358  4687        1729       1729  2015/2016     36  2016-04-30 00:00:00   
359  4688        1729       1729  2015/2016     36  2016-04-30 00:00:00   
360  4689        1729       1729  2015/2016     37  2016-05-07 00:00:00   
361  4690        1729       1729  2015/2016     37  2016-05-07 00:00:00   
362  4691        1729       1729  2015/2016     37  2016-05-07 00:00:00   
363  4692        1729       1729  2015/2016     37  2016-05-07 00:00:00   
364  4693        1729       1729  2015/2016     37  2016-05-08 00:00:00   
365  4694        1729       1729  2015/2016     37  2016-05-08 00:00:00   
366  4695        1729       1729  2015/2016     37  2016-05-07 00:00:00   
367  4696        1729       1729  2015/2016     37  2016-05-07 00:00:00   
368  4697        1729       1729  2015/2016     37  2016-05-08 00:00:00   
369  4698        1729       1729  2015/2016     37  2016-05-07 00:00:00   
370  4699        1729       1729  2015/2016     38  2016-05-15 00:00:00   
371  4700        1729       1729

In [14]:
# convert booleans into binary numbers 
simple_matches['home_wins'] = simple_matches['home_wins'].astype(int)
simple_matches['draws'] = simple_matches['draws'].astype(int)
simple_matches['away_wins'] = simple_matches['away_wins'].astype(int)


In [15]:
simple_matches.head()

id  country_id  league_id     season  stage                 date  \
0  4389        1729       1729  2015/2016      1  2015-08-09 00:00:00   
1  4390        1729       1729  2015/2016      1  2015-08-08 00:00:00   
2  4391        1729       1729  2015/2016      1  2015-08-08 00:00:00   
3  4392        1729       1729  2015/2016      1  2015-08-08 00:00:00   
4  4393        1729       1729  2015/2016      1  2015-08-08 00:00:00   

   match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
0       1987032              9825              8654               0   
1       1987033              8678             10252               0   
2       1987034              8455             10003               2   
3       1987035              8668              9817               2   
4       1987036              8197              8472               4   

   away_team_goal  home_goals  away_goals  home_wins  draws  away_wins  
0               2           0           2          0      0          1  
1               1           0           1          0      0          1  
2               2           2           2          0      1          0  
3               2           2           2          0      1          0  
4               2           4           2          1      0          0

In [16]:
# create cumulative home, draw and away wins grouping by each team id

simple_matches['cumulative_home_wins_home'] = simple_matches.groupby('home_team_api_id')['home_wins'].cumsum().fillna(0)
simple_matches['cumulative_away_wins_home'] = simple_matches.groupby('home_team_api_id')['home_wins'].cumsum().fillna(0)
simple_matches['cumulative_home_wins_away'] = simple_matches.groupby('away_team_api_id')['away_wins'].cumsum().fillna(0)
# simple_matches['cumulative_draws'] = simple_matches.groupby('home_team_api_id')['draws'].cumsum().fillna(0)
simple_matches['cumulative_away_wins_away'] = simple_matches.groupby('away_team_api_id')['away_wins'].cumsum().fillna(0)


In [17]:
simple_matches.tail(20)

id  country_id  league_id     season  stage                 date  \
360  4689        1729       1729  2015/2016     37  2016-05-07 00:00:00   
361  4690        1729       1729  2015/2016     37  2016-05-07 00:00:00   
362  4691        1729       1729  2015/2016     37  2016-05-07 00:00:00   
363  4692        1729       1729  2015/2016     37  2016-05-07 00:00:00   
364  4693        1729       1729  2015/2016     37  2016-05-08 00:00:00   
365  4694        1729       1729  2015/2016     37  2016-05-08 00:00:00   
366  4695        1729       1729  2015/2016     37  2016-05-07 00:00:00   
367  4696        1729       1729  2015/2016     37  2016-05-07 00:00:00   
368  4697        1729       1729  2015/2016     37  2016-05-08 00:00:00   
369  4698        1729       1729  2015/2016     37  2016-05-07 00:00:00   
370  4699        1729       1729  2015/2016     38  2016-05-15 00:00:00   
371  4700        1729       1729  2015/2016     38  2016-05-15 00:00:00   
372  4701        1729       1729  2015/2016     38  2016-05-15 00:00:00   
373  4702        1729       1729  2015/2016     38  2016-05-17 00:00:00   
374  4703        1729       1729  2015/2016     38  2016-05-15 00:00:00   
375  4704        1729       1729  2015/2016     38  2016-05-15 00:00:00   
376  4705        1729       1729  2015/2016     38  2016-05-15 00:00:00   
377  4706        1729       1729  2015/2016     38  2016-05-15 00:00:00   
378  4707        1729       1729  2015/2016     38  2016-05-15 00:00:00   
379  4708        1729       1729  2015/2016     38  2016-05-15 00:00:00   

     match_api_id  home_team_api_id  away_team_api_id  home_team_goal  \
360       1989070             10252             10261               0   
361       1989071              8678              8659               1   
362       1989072              9826             10194               2   
363       1989073              8197              8668               3   
364       1989074              8650              9817               2   
365       1989075              8456              9825               2   
366       1989076              9850             10260               0   
367       1989077              8472              8455               3   
368       1989078              8586              8466               1   
369       1989079              8654             10003               1   
370       1987597              9825             10252               4   
371       1987598              8455              8197               1   
372       1987599              8668              9850               3   
373       1987600             10260              8678               3   
374       1987601             10261              8586               5   
375       1987602              8466              9826               4   
376       1987603             10194              8654               2   
377       1987604             10003              8456               1   
378       1987605              9817              8472               2   
379       1987606              8659              8650               1   

     away_team_goal  home_goals  away_goals  home_wins  draws  away_wins  \
360               0          14          12          0      1          0   
361               1          23          14          0      1          0   
362               1          19          19          1      0          0   
363               1          35          24          1      0          0   
364               0          33          20          1      0          0   
365               2          47          34          0      1          0   
366               1          26          22          0      0          1   
367               2          23          27          1      0          0   
368               2          35          20          0      0          1   
369               4          34          22          0      0          1   
370               0          31          13          1      0

## Day two - Comparing bets (2015/2016)

The goal of today is to find a baseline betting profit and loss for our model to compare with.

From the data set we are going to look at the odds offered by the betting firm Ladbrokes on home win, away win and draw.

- Placing a ten pound bet on the favourite (team with lowest odds).
- Placing a ten pound bet on the Home team every time.
- Placing a ten pound bet on a draw every match.
- placing a ten pound bet on the underdog (team with highest odds).
- placing a ten pound bet on the away team.

Maximum loss is the ten pound stake - so after 380 matches worst case scenario is **£3800 down**. 


### Import the betting data

In [18]:
import pandas as pd
import numpy as np 
import sqlite3

In [19]:
# create the file path (include two beatthebookies)
path = '/Users/georgebrockman/code/georgebrockman/beatthebookies/beatthebookies/data/database.sqlite'
# create the connection
conn = sqlite3.connect(path)
# create the cursor
cursor = conn.cursor()

In [20]:
matches = pd.read_sql("""SELECT m.id, 
                        m.season, m.stage, m.date, 
                        ht.team_long_name as home_team, at.team_long_name as away_team, m.home_team_goal, 
                        m.away_team_goal, LBH, LBD, LBA                                      
                        FROM Match as m
                        LEFT JOIN Team AS ht on ht.team_api_id = m.home_team_api_id
                        LEFT JOIN Team AS at on at.team_api_id = m.away_team_api_id
                        WHERE league_id = 1729 AND season = '2015/2016'
                        ;""", conn)

In [21]:
# sort matches in to cronological order
matches.sort_values('date', inplace=True)

In [22]:
# check correctly imported the required data
matches.head()

id     season  stage                 date          home_team  \
1  4390  2015/2016      1  2015-08-08 00:00:00        Bournemouth   
2  4391  2015/2016      1  2015-08-08 00:00:00            Chelsea   
3  4392  2015/2016      1  2015-08-08 00:00:00            Everton   
4  4393  2015/2016      1  2015-08-08 00:00:00     Leicester City   
5  4394  2015/2016      1  2015-08-08 00:00:00  Manchester United   

           away_team  home_team_goal  away_team_goal   LBH  LBD   LBA  
1        Aston Villa               0               1  2.05  3.3   4.0  
2       Swansea City               2               2  1.40  4.5  10.0  
3            Watford               2               2  1.75  3.8   5.0  
4         Sunderland               4               2  2.00  3.4   4.2  
5  Tottenham Hotspur               1               0  1.67  4.0   5.5

In [23]:
# create booleans
matches['home_win'] = matches.home_team_goal > matches.away_team_goal
matches['draw'] = matches.home_team_goal == matches.away_team_goal
matches['away_win'] = matches.home_team_goal < matches.away_team_goal

In [24]:
# convert booleans into binary numbers 
matches['home_win'] = matches['home_win'].astype(int)
matches['draw'] = matches['draw'].astype(int)
matches['away_win'] = matches['away_win'].astype(int)

### Profit from betting on favourite

In [25]:
matches['fav_profit'] = -10

In [26]:
matches.head()

id     season  stage                 date          home_team  \
1  4390  2015/2016      1  2015-08-08 00:00:00        Bournemouth   
2  4391  2015/2016      1  2015-08-08 00:00:00            Chelsea   
3  4392  2015/2016      1  2015-08-08 00:00:00            Everton   
4  4393  2015/2016      1  2015-08-08 00:00:00     Leicester City   
5  4394  2015/2016      1  2015-08-08 00:00:00  Manchester United   

           away_team  home_team_goal  away_team_goal   LBH  LBD   LBA  \
1        Aston Villa               0               1  2.05  3.3   4.0   
2       Swansea City               2               2  1.40  4.5  10.0   
3            Watford               2               2  1.75  3.8   5.0   
4         Sunderland               4               2  2.00  3.4   4.2   
5  Tottenham Hotspur               1               0  1.67  4.0   5.5   

   home_win  draw  away_win  fav_profit  
1         0     0         1         -10  
2         0     1         0         -10  
3         0     1         0         -10  
4         1     0         0         -10  
5         1     0         0         -10

In [29]:
matches.loc[(matches[['LBH','LBD','LBA']].min(axis=1) == matches['LBH']) & (matches['home_win'] == 1), 'fav_profit'] = matches['LBH'] * 10
matches.loc[(matches[['LBH','LBD','LBA']].min(axis=1) == matches['LBD']) & (matches['draw'] == 1), 'fav_profit'] = matches['LBD'] * 10
matches.loc[(matches[['LBH','LBD','LBA']].min(axis=1) == matches['LBA']) & (matches['away_win'] == 1), 'fav_profit'] = matches['LBA'] * 10


In [30]:
matches

id     season  stage                 date             home_team  \
1    4390  2015/2016      1  2015-08-08 00:00:00           Bournemouth   
2    4391  2015/2016      1  2015-08-08 00:00:00               Chelsea   
3    4392  2015/2016      1  2015-08-08 00:00:00               Everton   
4    4393  2015/2016      1  2015-08-08 00:00:00        Leicester City   
5    4394  2015/2016      1  2015-08-08 00:00:00     Manchester United   
7    4396  2015/2016      1  2015-08-08 00:00:00          Norwich City   
0    4389  2015/2016      1  2015-08-09 00:00:00               Arsenal   
6    4395  2015/2016      1  2015-08-09 00:00:00      Newcastle United   
8    4397  2015/2016      1  2015-08-09 00:00:00            Stoke City   
9    4398  2015/2016      1  2015-08-10 00:00:00  West Bromwich Albion   
110  4499  2015/2016      2  2015-08-14 00:00:00           Aston Villa   
114  4503  2015/2016      2  2015-08-15 00:00:00           Southampton   
115  4504  2015/2016      2  2015-08-15 00:00:00            Sunderland   
116  4505  2015/2016      2  2015-08-15 00:00:00          Swansea City   
117  4506  2015/2016      2  2015-08-15 00:00:00     Tottenham Hotspur   
118  4507  2015/2016      2  2015-08-15 00:00:00               Watford   
119  4508  2015/2016      2  2015-08-15 00:00:00       West Ham United   
111  4500  2015/2016      2  2015-08-16 00:00:00        Crystal Palace   
113  4502  2015/2016      2  2015-08-16 00:00:00       Manchester City   
112  4501  2015/2016      2  2015-08-17 00:00:00             Liverpool   
229  4618  2015/2016      3  2015-08-22 00:00:00       West Ham United   
225  4614  2015/2016      3  2015-08-22 00:00:00          Norwich City   
224  4613  2015/2016      3  2015-08-22 00:00:00     Manchester United   
226  4615  2015/2016      3  2015-08-22 00:00:00            Sunderland   
221  4610  2015/2016      3  2015-08-22 00:00:00        Crystal Palace   
223  4612  2015/2016      3  2015-08-22 00:00:00        Leicester City   
222  4611  2015/2016      3  2015-08-23 00:00:00               Everton   
227  4616  2015/2016      3  2015-08-23 00:00:00               Watford   
228  4617  2015/2016      3  2015-08-23 00:00:00  West Bromwich Albion   
220  4609  2015/2016      3  2015-08-24 00:00:00               Arsenal   
..    ...        ...    ...                  ...                   ...   
290  4679  2015/2016     36  2016-04-30 00:00:00               Arsenal   
294  4683  2015/2016     36  2016-04-30 00:00:00      Newcastle United   
295  4684  2015/2016     36  2016-05-01 00:00:00           Southampton   
293  4682  2015/2016     36  2016-05-01 00:00:00     Manchester United   
297  4686  2015/2016     36  2016-05-01 00:00:00          Swansea City   
291  4680  2015/2016     36  2016-05-02 00:00:00               Chelsea   
300  4689  2015/2016     37  2016-05-07 00:00:00           Aston Villa   
301  4690  2015/2016     37  2016-05-07 00:00:00           Bournemouth   
302  4691  2015/2016     37  2016-05-07 00:00:00        Crystal Palace   
303  4692  2015/2016     37  2016-05-07 00:00:00        Leicester City   
309  4698  2015/2016     37  2016-05-07 00:00:00       West Ham United   
307  4696  2015/2016     37  2016-05-07 00:00:00            Sunderland   
306  4695  2015/2016     37  2016-05-07 00:00:00          Norwich City   
305  4694  2015/2016     37  2016-05-08 00:00:00       Manchester City   
308  4697  2015/2016     37  2016-05-08 00:00:00     Tottenham Hotspur   
304  4693  2015/2016     37  2016-05-08 00:00:00             Liverpool   
289  4678  2015/2016     35  2016-05-10 00:00:00       West Ham United   
286  4675  2015/2016     35  2016-05-11 00:00:00          Norwich City   
234  4623  2015/2016     30  2016-05-11 00:00:00             Liverpool   
238  4627  2015/2016     30  2016-05-11 00:00:00            Sunderland   
319  4708  2015/2016     38  2016-05-15 00:00:00  West Bromwich Albion   
318  4707  2015/2016     38  2016-05-15 00:00:00               Watford   
317  4706  

### Profit betting on Underdog

In [31]:
matches['dog_profit'] = -10

In [32]:
matches.loc[(matches[['LBH','LBD','LBA']].max(axis=1) == matches['LBH']) & (matches['home_win'] == 1), 'dog_profit'] = matches['LBH'] * 10
matches.loc[(matches[['LBH','LBD','LBA']].max(axis=1) == matches['LBD']) & (matches['draw'] == 1), 'dog_profit'] = matches['LBD'] * 10
matches.loc[(matches[['LBH','LBD','LBA']].max(axis=1) == matches['LBA']) & (matches['away_win'] == 1), 'dog_profit'] = matches['LBA'] * 10

In [33]:
matches.head()

id     season  stage                 date          home_team  \
1  4390  2015/2016      1  2015-08-08 00:00:00        Bournemouth   
2  4391  2015/2016      1  2015-08-08 00:00:00            Chelsea   
3  4392  2015/2016      1  2015-08-08 00:00:00            Everton   
4  4393  2015/2016      1  2015-08-08 00:00:00     Leicester City   
5  4394  2015/2016      1  2015-08-08 00:00:00  Manchester United   

           away_team  home_team_goal  away_team_goal   LBH  LBD   LBA  \
1        Aston Villa               0               1  2.05  3.3   4.0   
2       Swansea City               2               2  1.40  4.5  10.0   
3            Watford               2               2  1.75  3.8   5.0   
4         Sunderland               4               2  2.00  3.4   4.2   
5  Tottenham Hotspur               1               0  1.67  4.0   5.5   

   home_win  draw  away_win  fav_profit  dog_profit  
1         0     0         1       -10.0        40.0  
2         0     1         0       -10.0       -10.0  
3         0     1         0       -10.0       -10.0  
4         1     0         0        20.0       -10.0  
5         1     0         0        16.7       -10.0

### Profit from Home / Draw / Away

In [34]:
matches['home_profit'] = -10

In [35]:
matches.loc[(matches['home_win'] == 1), 'home_profit'] = matches['LBH'] * 10 

In [37]:
matches['draw_profit'] = -10

In [38]:
matches.loc[(matches['draw'] == 1), 'draw_profit'] = matches['LBD'] * 10 

In [39]:
matches['away_profit'] = -10

In [40]:
matches.loc[(matches['away_win'] == 1), 'away_profit'] = matches['LBA'] * 10 

In [70]:
matches.sort_values('date', inplace=True)

In [75]:
matches

id     season  stage                 date             home_team  \
1    4390  2015/2016      1  2015-08-08 00:00:00           Bournemouth   
2    4391  2015/2016      1  2015-08-08 00:00:00               Chelsea   
3    4392  2015/2016      1  2015-08-08 00:00:00               Everton   
4    4393  2015/2016      1  2015-08-08 00:00:00        Leicester City   
5    4394  2015/2016      1  2015-08-08 00:00:00     Manchester United   
7    4396  2015/2016      1  2015-08-08 00:00:00          Norwich City   
0    4389  2015/2016      1  2015-08-09 00:00:00               Arsenal   
6    4395  2015/2016      1  2015-08-09 00:00:00      Newcastle United   
8    4397  2015/2016      1  2015-08-09 00:00:00            Stoke City   
9    4398  2015/2016      1  2015-08-10 00:00:00  West Bromwich Albion   
110  4499  2015/2016      2  2015-08-14 00:00:00           Aston Villa   
116  4505  2015/2016      2  2015-08-15 00:00:00          Swansea City   
115  4504  2015/2016      2  2015-08-15 00:00:00            Sunderland   
114  4503  2015/2016      2  2015-08-15 00:00:00           Southampton   
119  4508  2015/2016      2  2015-08-15 00:00:00       West Ham United   
118  4507  2015/2016      2  2015-08-15 00:00:00               Watford   
117  4506  2015/2016      2  2015-08-15 00:00:00     Tottenham Hotspur   
111  4500  2015/2016      2  2015-08-16 00:00:00        Crystal Palace   
113  4502  2015/2016      2  2015-08-16 00:00:00       Manchester City   
112  4501  2015/2016      2  2015-08-17 00:00:00             Liverpool   
229  4618  2015/2016      3  2015-08-22 00:00:00       West Ham United   
224  4613  2015/2016      3  2015-08-22 00:00:00     Manchester United   
225  4614  2015/2016      3  2015-08-22 00:00:00          Norwich City   
223  4612  2015/2016      3  2015-08-22 00:00:00        Leicester City   
221  4610  2015/2016      3  2015-08-22 00:00:00        Crystal Palace   
226  4615  2015/2016      3  2015-08-22 00:00:00            Sunderland   
222  4611  2015/2016      3  2015-08-23 00:00:00               Everton   
227  4616  2015/2016      3  2015-08-23 00:00:00               Watford   
228  4617  2015/2016      3  2015-08-23 00:00:00  West Bromwich Albion   
220  4609  2015/2016      3  2015-08-24 00:00:00               Arsenal   
..    ...        ...    ...                  ...                   ...   
290  4679  2015/2016     36  2016-04-30 00:00:00               Arsenal   
294  4683  2015/2016     36  2016-04-30 00:00:00      Newcastle United   
295  4684  2015/2016     36  2016-05-01 00:00:00           Southampton   
293  4682  2015/2016     36  2016-05-01 00:00:00     Manchester United   
297  4686  2015/2016     36  2016-05-01 00:00:00          Swansea City   
291  4680  2015/2016     36  2016-05-02 00:00:00               Chelsea   
300  4689  2015/2016     37  2016-05-07 00:00:00           Aston Villa   
301  4690  2015/2016     37  2016-05-07 00:00:00           Bournemouth   
302  4691  2015/2016     37  2016-05-07 00:00:00        Crystal Palace   
303  4692  2015/2016     37  2016-05-07 00:00:00        Leicester City   
309  4698  2015/2016     37  2016-05-07 00:00:00       West Ham United   
307  4696  2015/2016     37  2016-05-07 00:00:00            Sunderland   
306  4695  2015/2016     37  2016-05-07 00:00:00          Norwich City   
305  4694  2015/2016     37  2016-05-08 00:00:00       Manchester City   
308  4697  2015/2016     37  2016-05-08 00:00:00     Tottenham Hotspur   
304  4693  2015/2016     37  2016-05-08 00:00:00             Liverpool   
289  4678  2015/2016     35  2016-05-10 00:00:00       West Ham United   
286  4675  2015/2016     35  2016-05-11 00:00:00          Norwich City   
234  4623  2015/2016     30  2016-05-11 00:00:00             Liverpool   
238  4627  2015/2016     30  2016-05-11 00:00:00            Sunderland   
319  4708  2015/2016     38  2016-05-15 00:00:00  West Bromwich Albion   
318  4707  2015/2016     38  2016-05-15 00:00:00               Watford   
317  4706  

### Cumulative profit totals for each strategey

In [85]:
# define a function that returns the maximum profit from each strategy
def c_profits(X):
    fav_profit_cum = X['fav_profit'].cumsum().max()
    dog_profit_cum = X['dog_profit'].cumsum().max()
    home_profit_cum = X['home_profit'].cumsum().max()
    draw_profit_cum = X['draw_profit'].cumsum().max()
    away_profit_cum = X['away_profit'].cumsum().max()
    print(f"if you chose the favourite strategy, your total profit is £{fav_profit_cum}; the underdog approach: £{dog_profit_cum}; home approach: £{home_profit_cum}; draw method: £{draw_profit_cum}; away tactic: £{away_profit_cum}")

In [86]:
c_profits(matches)

if you chose the favourite strategy, your total profit is £1426.0000000000002; the underdog approach: £1766.8; home approach: £1155.6; draw method: £1157.8; away tactic: £1591.9
